File: Graph Algorithm/cpp/counting_rooms.cpp
/*
    Author: JN.Wang
    Created: 03/07/24
*/

#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using pll = pair<ll, ll>;

/*
    Counting Rooms
    https://cses.fi/problemset/task/1192
*/

vector<pll> neighbors(vector<string> &mapa, pll position, ll rows, ll cols){
    vector<int> dx = {0, 0, 1, -1};
    vector<int> dy = {1, -1, 0, 0};
    vector<pll> nbs;

    for (int i = 0; i < 4; i++){
        ll newX = position.first + dx[i];
        ll newY = position.second + dy[i];
        
        if (newX >= 0 && newX < cols && newY >= 0 && newY < rows){
            if (mapa[newY][newX] != '#'){
                nbs.push_back(make_pair(newX, newY));
            }
        }
    }

    return nbs;
}

void dfs(vector<string> &mapa, pll position, vector<vector<int>> &vis, ll rows, ll cols){
    stack<pll> line;
    line.push(position);

    while (!line.empty()) {
        pll pos = line.top();
        line.pop();

        if (!vis[pos.second][pos.first]){
            vis[pos.second][pos.first] = 1;
            for (pll nb : neighbors(mapa, pos, rows, cols)){
                if(!vis[nb.second][nb.first])
                    line.push(nb);
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    ll n, m; cin >> n >> m;
    vector<string> mapa(n);
    for (auto &row : mapa)
        cin >> row;

    vector<vector<int>> vis(n, vector<int>(m));
    ll rooms = 0;
    for (int i = 0; i < n; i++){
        for (int j = 0; j < m; j++){
            if (mapa[i][j] != '#' && !vis[i][j]){
                dfs(mapa, make_pair(j, i), vis, n, m);
                ++rooms;
            }
        }
    }
    cout << rooms;
}

File: Graph Algorithm/python/counting_rooms.py
"""
    Author: JN.Wang
    Created: 03/07/24
"""

from sys import stdin
cin = stdin.readline

'''
    Counting Rooms
    https://cses.fi/problemset/task/1192

    Description:
        >   You are given a map of a building, 
            and your task is to count the number of its rooms. 
            
        >   The size of the map is n x m squares, 
            and each square is either floor or wall. 
            
        >   You can walk left, right, up, and down through the floor squares.

'''

from collections import deque

def neighbor(position, mapa):
    x, y = position
    nbs = []
    
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    for _ in range(4):
        new_x = x + dx[_]
        new_y = y + dy[_]
        
        if 0 <= new_x < m and 0 <= new_y < n:
            if mapa[new_y][new_x] != '#':
                nbs.append((new_x, new_y))
        
    return nbs

def dfs(pos, mapa):
    stack = deque()
    stack.append(pos)
    
    while stack:
        x, y = stack.pop()
        if not visited[y][x]:
            visited[y][x] = True
            for xx, yy in neighbor(position=(x, y), mapa=mapa):
                if not visited[yy][xx]:
                    stack.append((xx, yy))

def main():
    global n, m, visited
    n, m = map(int, input().split())
    mapa = [input().strip() for _ in range(n)]
    visited = [[False for j in range(m)] for i in range(n)]
    
    count = 0
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and mapa[i][j] != '#':
                pos = (j, i)
                dfs(pos=pos, mapa=mapa)
                count += 1
    
    print(count)
        
if __name__ == "__main__":
    main()
    